plugins/out_stackdriver/stackdriver.c

```diff
@@ -323,22 +327,68 @@ static int cb_stackdriver_init(struct flb_output_instance *ins,
     if (!token) {
         flb_plg_warn(ctx->ins, "token retrieval failed");
     }
-    if (ctx->metadata_server_auth) {
+
+    label = flb_kv_get_key_value("project_id", &ctx->resource_labels);
+    if (label == NULL) {
         ret = gce_metadata_read_project_id(ctx);
         if (ret == -1) {
+            flb_plg_error(ctx->ins, "get project_id failed");
             return -1;
         }
+    }
 
-        ret = gce_metadata_read_zone(ctx);
-        if (ret == -1) {
+    flb_plg_info(ctx->ins, "%s resource found", ctx->resource);
+
+    if (strncasecmp(ctx->resource, "gce_instance", 12) == 0) {
+        label = flb_kv_get_key_value("zone", &ctx->resource_labels);
+        if (label == NULL) {
+            ret = gce_metadata_read_zone(ctx);
+            if (ret == -1) {
+                flb_plg_error(ctx->ins, "get zone for gce_instance failed");
+                return -1;
+            }
+        }
+
+        label = flb_kv_get_key_value("instance_id", &ctx->resource_labels);
+        if (label == NULL) {
+            ret = gce_metadata_read_instance_id(ctx);
+            if (ret == -1) {
+                flb_plg_error(ctx->ins, "get instace_id for gce_instance failed");
+                return -1;
+            }
+        }
+    } else if (strncasecmp(ctx->resource, "generic_", 8) == 0) {
+        label = flb_kv_get_key_value("location", &ctx->resource_labels);
+        if (label == NULL) {
+            flb_plg_error(ctx->ins, "get location for generic_node/task failed");
             return -1;
         }
 
-        ret = gce_metadata_read_instance_id(ctx);
-        if (ret == -1) {
+        label = flb_kv_get_key_value("namespace", &ctx->resource_labels);
+        if (label == NULL) {
+            flb_plg_error(ctx->ins, "get namespace for generic_node/task failed");
             return -1;
         }
 
+        if (strncasecmp(ctx->resource, "generic_node", 12) == 0) {;
+            label = flb_kv_get_key_value("node_id", &ctx->resource_labels);
+            if (label == NULL) {
+                flb_plg_error(ctx->ins, "get node_id for generic_node failed");
+                return -1;
+            }
+        } else {
+            label = flb_kv_get_key_value("job", &ctx->resource_labels);
+            if (label == NULL) {
+                flb_plg_error(ctx->ins, "get job for generic_task failed");
+                return -1;
+            }
+
+            label = flb_kv_get_key_value("task_id", &ctx->resource_labels);
+            if (label == NULL) {
+                flb_plg_error(ctx->ins, "get task_id for generic_task failed");
+                return -1;
+            }
+        }
```

Comment:
Hmm, it looks like you're hard-coding the resource schema (all of the required labels) in the agent code. That will require keeping that schema up-to-date in the future. Why not just send what the user specified in the config and let the API reject your request if some labels are missing?

Reply:
So this is the way it was previously, but that doesn't mean its the best. The one benefit of doing it this way is the immediate feedback that a label was wrong. I don't have any real problem doing it without encoding the schema, the only thing I am concerned with is the rejection propagating back to the user.

Reply:
So I still end up hard coding the schema in a way, but it simplifies how to create the resource. Since non-expected resource labels get silently dropped, I don't see a problem here except for potential divergence further down the road.